vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ModuleLinkManager.java

```diff
@@ -0,0 +1,125 @@
+package edu.asu.diging.vspace.core.services.impl;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import javax.transaction.Transactional;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import edu.asu.diging.vspace.core.data.ModuleLinkRepository;
+import edu.asu.diging.vspace.core.data.display.ModuleLinkDisplayRepository;
+import edu.asu.diging.vspace.core.exception.ImageCouldNotBeStoredException;
+import edu.asu.diging.vspace.core.exception.LinkDoesNotExistsException;
+import edu.asu.diging.vspace.core.exception.SpaceDoesNotExistException;
+import edu.asu.diging.vspace.core.factory.IModuleLinkDisplayFactory;
+import edu.asu.diging.vspace.core.factory.IModuleLinkFactory;
+import edu.asu.diging.vspace.core.model.IModule;
+import edu.asu.diging.vspace.core.model.IModuleLink;
+import edu.asu.diging.vspace.core.model.ISpace;
+import edu.asu.diging.vspace.core.model.display.DisplayType;
+import edu.asu.diging.vspace.core.model.display.ILinkDisplay;
+import edu.asu.diging.vspace.core.model.display.IModuleLinkDisplay;
+import edu.asu.diging.vspace.core.model.display.impl.ModuleLinkDisplay;
+import edu.asu.diging.vspace.core.model.impl.ModuleLink;
+import edu.asu.diging.vspace.core.services.IModuleLinkManager;
+import edu.asu.diging.vspace.core.services.IModuleManager;
+import edu.asu.diging.vspace.core.services.ISpaceManager;
+
+@Transactional
+@Service
+public class ModuleLinkManager extends LinkManager implements IModuleLinkManager{
```

Comment:
I think the template pattern would work well here:
- LinkManager has delete/create/update/...Link methods that call abstract methods which are implemented in the subclasses and do the link specific stuff.
- create a link superclass/interface that has the common stuff (owning space)
- you can probably even use generics to make this more elegant, but maybe start with the other stuff first.

Reply:
I did not understand methods calling an abstract method part in the first bullet point. Currently, LinkManager contains abstract methods whose implementation is written in child classes like ExternalLinkManager and its interfaces are used to track implementation of which class is to be invoked while creating or deleting links. Can we please discuss this on zoom?

Reply:
Yes, we can discuss this after my meeting. But basically you would have something like this (pseudo code):
```
class LinkManager:
      method createLink():
             // get space to the link belongs to
             
             // call abstract method to get link target
             target = getTarget()
             
             // call abstract method to create link
             link = createLinkObject()
             ...

class ModuleLinkManager:
        method getTarget():
               // gets and returns the link target (for module links the linked module)
```